sdk/src/main/java/com/amplitude/experiment/DefaultExperimentClient.kt

```diff
@@ -367,17 +372,265 @@ internal class DefaultExperimentClient internal constructor(
             .library("experiment-android-client/${BuildConfig.VERSION_NAME}")
             .build().merge(providedUser)
     }
+
+
+    private fun addContext(user: ExperimentUser?): ExperimentUser {
+        val providedUser = this.config.userProvider?.getUser()
+        val mergedUserProperties = (user?.userProperties ?: emptyMap()) + (providedUser?.userProperties ?: emptyMap())
+
+        return ExperimentUser.builder()
+            .library("experiment-android-client/$VERSION_NAME")
+            .merge(this.config.userProvider?.getUser()).merge(this.user)
+            .userProperties(mergedUserProperties)
+            .build()
+    }
+
+    private fun evaluate(flagKeys: Set<String>): Map<String, Variant> {
+        val user = addContext(this.user)
+        val flags = topologicalSort(this.flags.getAll(), flagKeys)
+        val context = EvaluationContext().apply { put("user", user) }
```

Comment:
We need to apply groups and group properties differently see here for implementation:

This is something I missed in the original javascript implementation: https://github.com/amplitude/experiment-ios-client/blob/19abd315fbc7a416d1dc8ca92950ef2e5c710c29/Sources/Experiment/ExperimentUser.swift#L585